File: mini-agent-harness/typescript/h0.ts
export type HarnessMessage =
  | { kind: 'user'; text: string }
  | { kind: 'assistant'; text: string }
  | { kind: 'progress'; text: string }

export type RunState =
  | { status: 'idle' }
  | { status: 'running' }
  | { status: 'completed' }
  | { status: 'failed'; error: string }
  | { status: 'cancelled'; reason: string }

export type H0EventInput =
  | { type: 'run.state'; status: RunState['status'] }
  | { type: 'call.entered'; source: string; target: string }
  | { type: 'message.appended'; owner: 'H0Harness'; kind: HarnessMessage['kind']; size: number }
  | { type: 'view.snapshotted'; owner: 'H0Harness'; size: number }
  | { type: 'query.event'; kind: HarnessMessage['kind'] }
  | { type: 'branch.skipped'; branch: 'query'; reason: string }
  | { type: 'cancel.requested'; reason: string }
  | { type: 'run.failed'; message: string }
  | { type: 'cleanup.finished' }

export type H0Event = H0EventInput & { seq: number }

export class TraceLog {
  readonly events: H0Event[] = []
  #nextSeq = 1

  record(event: H0EventInput): void {
    this.events.push({ ...event, seq: this.#nextSeq++ } as H0Event)
  }
}

const allowedTransitions: Record<RunState['status'], RunState['status'][]> = {
  idle: ['running'],
  running: ['completed', 'failed', 'cancelled'],
  completed: [],
  failed: [],
  cancelled: [],
}

export function transition(_current: RunState, next: RunState): RunState {
  if (!allowedTransitions[_current.status].includes(next.status)) {
    throw new Error(`invalid transition: ${_current.status} -> ${next.status}`)
  }
  return next
}

export function parseMessage(value: unknown): HarnessMessage {
  if (typeof value !== 'object' || value === null) {
    throw new Error('message must be an object')
  }
  const candidate = value as Record<string, unknown>
  if (
    (candidate.kind === 'user' ||
      candidate.kind === 'assistant' ||
      candidate.kind === 'progress') &&
    typeof candidate.text === 'string'
  ) {
    return { kind: candidate.kind, text: candidate.text }
  }
  throw new Error('invalid message')
}

export class ResourceScope {
  #disposers: Array<() => void | Promise<void>> = []
  #disposed = false

  register(disposer: () => void | Promise<void>): void {
    if (this.#disposed) throw new Error('resource scope already disposed')
    this.#disposers.push(disposer)
  }

  async dispose(): Promise<void> {
    if (this.#disposed) return
    this.#disposed = true
    for (const disposer of this.#disposers.reverse()) await disposer()
    this.#disposers = []
  }
}

export class CancellationScope {
  readonly #controller = new AbortController()
  readonly #parent: AbortSignal | undefined
  readonly #forwardParent: () => void
  #timer: ReturnType<typeof setTimeout> | undefined

  constructor(parent?: AbortSignal, timeoutMs?: number) {
    this.#parent = parent
    this.#forwardParent = () => this.cancel(String(parent?.reason ?? 'parent'))
    if (parent?.aborted) {
      this.#forwardParent()
    } else {
      parent?.addEventListener('abort', this.#forwardParent, { once: true })
      if (timeoutMs !== undefined) {
        this.#timer = setTimeout(() => this.cancel('timeout'), timeoutMs)
        this.#timer.unref?.()
      }
    }
  }

  get signal(): AbortSignal {
    return this.#controller.signal
  }

  cancel(reason: string): void {
    if (!this.signal.aborted) this.#controller.abort(reason)
  }

  cleanup(): void {
    if (this.#timer !== undefined) clearTimeout(this.#timer)
    this.#parent?.removeEventListener('abort', this.#forwardParent)
  }
}

export type ProcessInputResult = {
  messages: HarnessMessage[]
  shouldQuery: boolean
}

export type ProcessInput = (
  prompt: string,
  history: readonly HarnessMessage[],
) => Promise<ProcessInputResult>

export type QueryStream = (
  messages: readonly HarnessMessage[],
  signal: AbortSignal,
) => AsyncIterable<HarnessMessage>

export class H0Harness {
  #messages: HarnessMessage[] = []
  #state: RunState = { status: 'idle' }
  readonly #processInput: ProcessInput
  readonly #queryStream: QueryStream
  readonly trace = new TraceLog()

  constructor(processInput: ProcessInput, queryStream: QueryStream) {
    this.#processInput = processInput
    this.#queryStream = queryStream
  }

  get state(): RunState {
    return this.#state
  }

  get messages(): readonly HarnessMessage[] {
    return [...this.#messages]
  }

  #setState(next: RunState): void {
    this.#state = transition(this.#state, next)
    this.trace.record({ type: 'run.state', status: next.status })
  }

  async *run(
    prompt: string,
    options: { parentSignal?: AbortSignal; timeoutMs?: number } = {},
  ): AsyncGenerator<HarnessMessage, void> {
    const cancellation = new CancellationScope(
      options.parentSignal,
      options.timeoutMs,
    )
    const resources = new ResourceScope()
    resources.register(() => cancellation.cleanup())
    const onAbort = (): void => {
      this.trace.record({
        type: 'cancel.requested',
        reason: String(cancellation.signal.reason ?? 'cancelled'),
      })
    }
    cancellation.signal.addEventListener('abort', onAbort, { once: true })
    resources.register(() => cancellation.signal.removeEventListener('abort', onAbort))

    this.#setState({ status: 'running' })
    try {
      this.trace.record({
        type: 'call.entered',
        source: 'H0Harness.run',
        target: 'processInput',
      })
      const processed = await this.#processInput(prompt, [...this.#messages])
      for (const message of processed.messages) {
        this.#messages.push(message)
        this.trace.record({
          type: 'message.appended',
          owner: 'H0Harness',
          kind: message.kind,
          size: this.#messages.length,
        })
      }
      const requestView = [...this.#messages]
      this.trace.record({
        type: 'view.snapshotted',
        owner: 'H0Harness',
        size: requestView.length,
      })

      if (cancellation.signal.aborted) {
        this.#setState({
          status: 'cancelled',
          reason: String(cancellation.signal.reason ?? 'cancelled'),
        })
        return
      }
      if (!processed.shouldQuery) {
        this.trace.record({
          type: 'branch.skipped',
          branch: 'query',
          reason: 'processInput.shouldQuery=false',
        })
        this.#setState({ status: 'completed' })
        return
      }

      this.trace.record({
        type: 'call.entered',
        source: 'H0Harness.run',
        target: 'queryStream',
      })
      for await (const message of this.#queryStream(
        requestView,
        cancellation.signal,
      )) {
        if (cancellation.signal.aborted) break
        this.trace.record({ type: 'query.event', kind: message.kind })
        this.#messages.push(message)
        this.trace.record({
          type: 'message.appended',
          owner: 'H0Harness',
          kind: message.kind,
          size: this.#messages.length,
        })
        yield message
      }

      if (cancellation.signal.aborted) {
        this.#setState({
          status: 'cancelled',
          reason: String(cancellation.signal.reason ?? 'cancelled'),
        })
      } else {
        this.#setState({ status: 'completed' })
      }
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error)
      this.trace.record({ type: 'run.failed', message })
      this.#setState({ status: 'failed', error: message })
      throw error
    } finally {
      await resources.dispose()
      this.trace.record({ type: 'cleanup.finished' })
    }
  }
}

export async function collect(
  stream: AsyncIterable<HarnessMessage>,
): Promise<HarnessMessage[]> {
  const messages: HarnessMessage[] = []
  for await (const message of stream) messages.push(message)
  return messages
}

File: mini-agent-harness/python/h0.py
from __future__ import annotations

import asyncio
from collections.abc import AsyncIterator, Awaitable, Callable, Sequence
from dataclasses import asdict, dataclass
from typing import Literal, TypedDict


class HarnessMessage(TypedDict):
    kind: Literal["user", "assistant", "progress"]
    text: str


@dataclass(frozen=True)
class RunState:
    status: Literal["idle", "running", "completed", "failed", "cancelled"]
    error: str = ""
    reason: str = ""


@dataclass(frozen=True)
class H0Event:
    seq: int
    type: str
    details: dict[str, object]


class TraceLog:
    def __init__(self) -> None:
        self.events: list[H0Event] = []

    def record(self, event_type: str, **details: object) -> None:
        self.events.append(H0Event(len(self.events) + 1, event_type, details))

    def to_dicts(self) -> list[dict[str, object]]:
        return [asdict(event) for event in self.events]


ALLOWED_TRANSITIONS = {
    "idle": {"running"},
    "running": {"completed", "failed", "cancelled"},
    "completed": set(),
    "failed": set(),
    "cancelled": set(),
}


def transition(current: RunState, next_state: RunState) -> RunState:
    if next_state.status not in ALLOWED_TRANSITIONS[current.status]:
        raise ValueError(f"invalid transition: {current.status} -> {next_state.status}")
    return next_state


def parse_message(value: object) -> HarnessMessage:
    if not isinstance(value, dict):
        raise ValueError("message must be an object")
    kind = value.get("kind")
    text = value.get("text")
    if kind not in {"user", "assistant", "progress"} or not isinstance(text, str):
        raise ValueError("invalid message")
    return {"kind": kind, "text": text}  # type: ignore[return-value]


class ResourceScope:
    def __init__(self) -> None:
        self._disposers: list[Callable[[], object | Awaitable[object]]] = []
        self._disposed = False

    def register(self, disposer: Callable[[], object | Awaitable[object]]) -> None:
        if self._disposed:
            raise RuntimeError("resource scope already disposed")
        self._disposers.append(disposer)

    async def dispose(self) -> None:
        if self._disposed:
            return
        self._disposed = True
        for disposer in reversed(self._disposers):
            result = disposer()
            if asyncio.iscoroutine(result):
                await result
        self._disposers.clear()


class CancellationScope:
    def __init__(self) -> None:
        self._event = asyncio.Event()
        self.reason = ""
        self._callbacks: list[Callable[[str], None]] = []

    @property
    def cancelled(self) -> bool:
        return self._event.is_set()

    def cancel(self, reason: str) -> None:
        if self.cancelled:
            return
        self.reason = reason
        self._event.set()
        for callback in tuple(self._callbacks):
            callback(reason)

    async def wait(self) -> None:
        await self._event.wait()

    def add_callback(self, callback: Callable[[str], None]) -> Callable[[], None]:
        self._callbacks.append(callback)
        if self.cancelled:
            callback(self.reason)

        def remove() -> None:
            if callback in self._callbacks:
                self._callbacks.remove(callback)

        return remove


@dataclass(frozen=True)
class ProcessInputResult:
    messages: list[HarnessMessage]
    should_query: bool


ProcessInput = Callable[
    [str, Sequence[HarnessMessage]], Awaitable[ProcessInputResult]
]
QueryStream = Callable[
    [Sequence[HarnessMessage], CancellationScope], AsyncIterator[HarnessMessage]
]


class H0Harness:
    def __init__(self, process_input: ProcessInput, query_stream: QueryStream) -> None:
        self._process_input = process_input
        self._query_stream = query_stream
        self._messages: list[HarnessMessage] = []
        self.state = RunState("idle")
        self.trace = TraceLog()

    @property
    def messages(self) -> list[HarnessMessage]:
        return list(self._messages)

    def _set_state(self, next_state: RunState) -> None:
        self.state = transition(self.state, next_state)
        self.trace.record("run.state", status=next_state.status)

    async def run(
        self, prompt: str, cancellation: CancellationScope | None = None
    ) -> AsyncIterator[HarnessMessage]:
        active_cancel = cancellation or CancellationScope()
        resources = ResourceScope()
        remove_callback = active_cancel.add_callback(
            lambda reason: self.trace.record("cancel.requested", reason=reason)
        )
        resources.register(remove_callback)

        self._set_state(RunState("running"))
        try:
            self.trace.record(
                "call.entered", source="H0Harness.run", target="processInput"
            )
            processed = await self._process_input(prompt, list(self._messages))
            for message in processed.messages:
                self._messages.append(message)
                self.trace.record(
                    "message.appended",
                    owner="H0Harness",
                    kind=message["kind"],
                    size=len(self._messages),
                )
            request_view = list(self._messages)
            self.trace.record(
                "view.snapshotted", owner="H0Harness", size=len(request_view)
            )

            if active_cancel.cancelled:
                self._set_state(RunState("cancelled", reason=active_cancel.reason))
                return
            if not processed.should_query:
                self.trace.record(
                    "branch.skipped",
                    branch="query",
                    reason="processInput.shouldQuery=false",
                )
                self._set_state(RunState("completed"))
                return

            self.trace.record(
                "call.entered", source="H0Harness.run", target="queryStream"
            )
            async for message in self._query_stream(request_view, active_cancel):
                if active_cancel.cancelled:
                    break
                self.trace.record("query.event", kind=message["kind"])
                self._messages.append(message)
                self.trace.record(
                    "message.appended",
                    owner="H0Harness",
                    kind=message["kind"],
                    size=len(self._messages),
                )
                yield message

            if active_cancel.cancelled:
                self._set_state(RunState("cancelled", reason=active_cancel.reason))
            else:
                self._set_state(RunState("completed"))
        except Exception as error:
            self.trace.record("run.failed", message=str(error))
            self._set_state(RunState("failed", error=str(error)))
            raise
        finally:
            await resources.dispose()
            self.trace.record("cleanup.finished")


async def collect(stream: AsyncIterator[HarnessMessage]) -> list[HarnessMessage]:
    return [message async for message in stream]
